src/persist-client/src/stats.rs

```diff
@@ -146,16 +146,37 @@ where
         }
     };
 
+    let mut k_buf: Option<K> = None;
+    let mut v_buf: Option<V> = None;
+
+    let mut k_storage = Some(K::Storage::default());
+    let mut v_storage = Some(V::Storage::default());
+
     if format.is_structured() {
         let mut builder = PartBuilder2::new(schemas.key.as_ref(), schemas.val.as_ref());
         for update in updates {
             for ((k, v), t, d) in update.iter() {
-                let k = K::decode(k)?;
-                let v = V::decode(v)?;
                 let t = i64::from_le_bytes(t);
                 let d = i64::from_le_bytes(d);
 
-                builder.push(&k, &v, t, d);
+                // Attempt to re-use allocations as much as possible.
+                match (k_buf.as_mut(), v_buf.as_mut()) {
+                    (Some(k_buf), Some(v_buf)) => {
+                        K::decode_from(k_buf, k, &mut k_storage)?;
+                        V::decode_from(v_buf, v, &mut v_storage)?;
```

Comment:
I found the `k_buf` / `v_buf` naming here confusing, since in `decode_from` the `&[u8]` is called `buf`... I can't think of anything I love, but maybe `k_cached` or `k_into`, or swap the names?

Reply:
Good call! I changed it to `k_reuse`, which seemed decent?